File: C#/TopflytechCodec/TopflytechCodec/Entities/Crypto.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Security.Cryptography;
using System.IO;

namespace TopflytechCodec.Entities
{
    class Crypto
    {
        private const String ivParam = "topflytech201205";
        public static byte[] AESEncrypt(byte[] toEncrypt, string key)
        {
            byte[] keyArray = System.Text.Encoding.Default.GetBytes(key);
            byte[] ivArray = System.Text.Encoding.Default.GetBytes(ivParam);
            byte[] toEncryptArray = toEncrypt;
            RijndaelManaged rDel = new RijndaelManaged();
            MD5 md5 = new MD5CryptoServiceProvider();
            byte[] output = md5.ComputeHash(keyArray);
            rDel.Key = output;
            rDel.IV = ivArray;
            rDel.Mode = CipherMode.CBC;
            rDel.Padding = PaddingMode.PKCS7;
            ICryptoTransform cTransform = rDel.CreateEncryptor();
            byte[] resultArray = cTransform.TransformFinalBlock(toEncryptArray, 0, toEncryptArray.Length);
            return resultArray;
        }
        public static byte[] AESDecrypt(byte[] toDecrypt, string key)
        {
            byte[] keyArray = System.Text.Encoding.Default.GetBytes(key);
            byte[] ivArray = System.Text.Encoding.Default.GetBytes(ivParam);
            byte[] toEncryptArray = toDecrypt;
            MD5 md5 = new MD5CryptoServiceProvider();
            byte[] output = md5.ComputeHash(keyArray);
            RijndaelManaged rDel = new RijndaelManaged();
            rDel.Key = output;
            rDel.IV = ivArray;
            rDel.Mode = CipherMode.CBC;
            rDel.Padding = PaddingMode.PKCS7;
            ICryptoTransform cTransform = rDel.CreateDecryptor();
            byte[] resultArray = cTransform.TransformFinalBlock(toEncryptArray, 0, toEncryptArray.Length);
            return resultArray;
        }


        public static byte[] MD5(byte[] data)
        { 
            RijndaelManaged rDel = new RijndaelManaged();
            MD5 md5 = new MD5CryptoServiceProvider();
            byte[] output = md5.ComputeHash(data);
            byte[] result = new byte[8];
            Array.Copy(output, 4, result, 0, 8);
            return result; 
        }

        public static int GetAesLength(int packageLength)
        {
            if (packageLength <= 15)
            {
                return packageLength;
            }
            return ((packageLength - 15) / 16 + 1) * 16 + 15;
        }


        public static byte[] DecryptData(byte[] data, int encryptType, string aesKey)
        {
            if (encryptType == MessageEncryptType.MD5)
            {
                byte[] realData = new byte[data.Length - 8];
                byte[] md5Data = new byte[8];
                Array.Copy(data, 0, realData, 0, data.Length - 8);
                Array.Copy(data, data.Length - 8, md5Data, 0, 8);
                byte[] pathMd5 = Crypto.MD5(realData);
                if (pathMd5 == null)
                {
                    return null;
                }
                if (!Utils.ArrayEquals(md5Data, pathMd5))
                {
                    return null;
                }
                return realData;
            }
            else if (encryptType == MessageEncryptType.AES)
            {
                if (data.Length <= 15)
                {
                    return data;
                }
                byte[] head = new byte[15];
                byte[] aesData = new byte[data.Length - 15];
                Array.Copy(data, 0, head, 0, 15);
                Array.Copy(data, 15, aesData, 0, data.Length - 15);
                byte[] realData = Crypto.AESDecrypt(aesData, aesKey);
                if (realData == null)
                {
                    return null;
                }
                MemoryStream memoryStream = new MemoryStream();
                try
                {
                    memoryStream.Write(head, 0, head.Length);
                    memoryStream.Write(realData, 0, realData.Length);
                    return memoryStream.ToArray();
                }
                finally
                {
                    memoryStream.Close();
                }
            }
            else
            {
                return data;
            }
        }
    }
}

File: Java/topflytechCodec/src/com/topflytech/codec/Crypto.java
package com.topflytech.codec;

import com.topflytech.codec.entities.MessageEncryptType;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.security.MessageDigest;
import java.util.Arrays;

/**
 * Created by admin on 2017/5/10.
 */
public class Crypto {
    public static byte[] MD5(byte[] data) {
        try {
            MessageDigest messageDigest = MessageDigest.getInstance("MD5");
            byte[] digest = messageDigest.digest(data);
            byte[] pathMd5 = Arrays.copyOfRange(digest, 4, 12);
            return pathMd5;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    private static String ivParam = "topflytech201205";

    public static byte[] aesEncrypt(byte[] src, String key) throws Exception {
        if (key == null) {
            System.out.print("Key is null");
            return null;
        }
        byte[] raw = key.getBytes();
        MessageDigest md = MessageDigest.getInstance("MD5");
        byte[] arrBTmp = md.digest(raw);
        SecretKeySpec skeySpec = new SecretKeySpec(arrBTmp, "AES");
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");

        IvParameterSpec iv = new IvParameterSpec(ivParam.getBytes());
        cipher.init(Cipher.ENCRYPT_MODE, skeySpec, iv);
        byte[] encrypted = cipher.doFinal(src);
        return encrypted;
    }

    public static byte[] aesDecrypt(byte[] src, String key) throws Exception {
        try {
            if (key == null) {
                System.out.print("Key为空null");
                return null;
            }
            byte[] raw = key.getBytes("ASCII");
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] arrBTmp = md.digest(raw);
            SecretKeySpec skeySpec = new SecretKeySpec(arrBTmp, "AES");
            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            IvParameterSpec iv = new IvParameterSpec(ivParam.getBytes());
            cipher.init(Cipher.DECRYPT_MODE, skeySpec, iv);
            try {
                byte[] original = cipher.doFinal(src);
                return original;
            } catch (Exception e) {
                System.out.println(e.toString());
                return null;
            }
        } catch (Exception ex) {
            System.out.println(ex.toString());
            return null;
        }
    }
    public static int getAesLength(int packageLength) {
        if (packageLength <= 15){
            return packageLength;
        }
        return ((packageLength - 15) / 16 + 1) * 16 + 15;
    }
    public static byte[] decryptData(byte[] data, int encryptType, String aesKey){
        if (encryptType == MessageEncryptType.MD5){
            byte[] realData = Arrays.copyOfRange(data, 0, data.length - 8);
            byte[] md5Data = Arrays.copyOfRange(data, data.length - 8, data.length);
            byte[] pathMd5 = Crypto.MD5(realData);
            if (pathMd5 == null){
                return null;
            }
            if (!Arrays.equals(pathMd5,md5Data)){
                return null;
            }
            return realData;
        }else if(encryptType == MessageEncryptType.AES){
            byte[] head = Arrays.copyOfRange(data,0,15);
            byte[] aesData = Arrays.copyOfRange(data,15,data.length);
            if (aesData == null || aesData.length == 0){
                return data;
            }
            byte[] realData = null;
            try {
                realData = Crypto.aesDecrypt(aesData, aesKey);
            } catch (Exception e) {
                e.printStackTrace();
            }
            if (realData == null){
                return null;
            }
            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
            try {
                try {
                    outputStream.write(head);
                    outputStream.write(realData);
                    return outputStream.toByteArray();
                } finally {
                    outputStream.close();
                }
            }catch (IOException e) {
                e.printStackTrace();
                return null;
            }
        }else {
            return data;
        }
    }
}
